dali/kernels/reduce/reduce_all_gpu_test.cu

```diff
@@ -0,0 +1,165 @@
+#include "dali/core/util.h"
+#include "dali/kernels/alloc.h"
+#include "dali/core/span.h"
+#include "dali/core/cuda_event.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename T>
+double RefSum(span<T> in) {
+  switch (in.size()) {
+    case 0:
+      return 0;
+    case 1:
+      return in[0];
+    default: {
+      if (in.size() <= 128) {
+        double acc = 0;
+        for (auto &x : in)
+          acc += x;
+        return acc;
+      }
+      int m = in.size() / 2;
+      int n = in.size() - m;
+      return RefSum(make_span(in.data(), m)) + RefSum(make_span(in.data() + m, n));
+    }
+  }
+}
+
+TEST(ReduceGPU, ReduceAllKernel) {
+  std::mt19937_64 rng(1234);
+  std::uniform_real_distribution<float> dist(0, 1);
+
+  int n_in = 1<<23;  // 8M numbers
+  dim3 block(32, 32);
+  int nblock = 1024;
+  int n_out0 = std::min<int>(div_ceil(n_in, nblock), 1024);
+  int n_out = n_out0 + 1;
+  auto in_data = memory::alloc_unique<float>(AllocType::GPU, n_in);
+  auto out_data = memory::alloc_unique<float>(AllocType::GPU, n_out);
+  std::vector<float> in_cpu(n_in), out_cpu(n_out);
+  for (auto &x : in_cpu)
+    x = dist(rng);
+  double ref_sum = RefSum(make_cspan(in_cpu));
+
+  cudaMemcpy(in_data.get(), in_cpu.data(), n_in * sizeof(*in_data), cudaMemcpyHostToDevice);
+
+  dim3 grid = n_out0;
+  ReduceAllKernel<<<1, block>>>(out_data.get(), in_data.get(), n_in);
+  cudaDeviceSynchronize();
+  auto start = CUDAEvent::CreateWithFlags(0);
+  auto end =   CUDAEvent::CreateWithFlags(0);
+  cudaEventRecord(start);
+  ReduceAllKernel<<<grid, block>>>(out_data.get() + 1, in_data.get(), n_in);
+  ReduceAllKernel<<<1, block>>>(out_data.get(), out_data.get() + 1, n_out0);
+  cudaEventRecord(end);
+  cudaMemcpy(out_cpu.data(), out_data.get(), n_out * sizeof(*out_data), cudaMemcpyDeviceToHost);
+  cudaDeviceSynchronize();
+  float t = 0;
+  cudaEventElapsedTime(&t, start, end);
+  double out_sum = out_cpu[0];
+  double out_partial = RefSum(make_cspan(&out_cpu[1], n_out0));
+  EXPECT_NEAR(out_sum, ref_sum, ref_sum * 1e-7 + 1e-7);
+  EXPECT_NEAR(out_partial, ref_sum, ref_sum * 1e-7 + 1e-7);
+
+  t /= 1000;
+  std::cout << n_in * sizeof(*in_data) / t * 1e-9 << " GB/s" << std::endl;
+}
+
+TEST(ReduceGPU, ReduceAllBatchedKernel) {
+  std::mt19937_64 rng(1234);
+  std::uniform_real_distribution<float> dist(0, 1);
+  std::uniform_int_distribution<int> size_dist(10000, 1000000);
+
+  int64_t n_in = 0;
+  int samples = 35;
+  std::vector<int64_t> sizes(samples);
+  for (auto &size : sizes) {
+    size = size_dist(rng);
+    n_in += size;
+  }
+
+  dim3 block(32, 32);
+  int n_out_per_sample = 32;
+  int n_out0 = samples * n_out_per_sample;
+  int n_out = n_out0 + samples;
+  dim3 grid(n_out_per_sample, samples);
+  auto in_data = memory::alloc_unique<float>(AllocType::GPU, n_in);
+  auto out_data = memory::alloc_unique<float>(AllocType::GPU, n_out);
+  std::vector<float> in_cpu(n_in), out_cpu(n_out);
+  for (auto &x : in_cpu)
+    x = dist(rng);
```

Comment:
the code generating the random data is duplicated. Can we extract it? Having a utility to generate random data and copy to the GPU would be handy when writing other tests